src/supervisor/models.py

```diff
@@ -0,0 +1,126 @@
+"""
+This module holds the classes used by supervisor
+"""
+
+import logging
+import os
+import sys
+import traceback
+
+from .message_handlers import ConsoleHandler, MessageHandler
+
+
+class Supervisor:
+    """Captures exceptions and sends these and other notifications via email or other methods.
+
+    Attributes
+    ----------
+    project_name : str
+        The name of the parent project using Supervisor; used to get it's logfile
+    message_handlers : [MessageHandler]
+        Notifications will be sent via all handlers in this list
+    log_path : Path
+        Path to the logfile
+
+    Methods
+    -------
+    add_message_handler(handler)
+        Register a new handler for sending messages
+    notify(message)
+        Send a notification to all registered handlers
+    _manage_exceptions
+        Closure around a replacement for sys.excepthook
+    """
+
+    def __init__(self, project_name, log_path=None):
+
+        #: Set up our list of MessageHandlers and add a default ConsoleHandler
+        self.project_name = project_name
+        self.message_handlers = []
+        self.message_handlers.append(ConsoleHandler())
+        self.log_path = log_path
+
+        #: Catch any uncaught exception with our custom exception handler
+        sys.excepthook = self._manage_exceptions()
+
+    def add_message_handler(self, handler: MessageHandler):
+        """Register a new handler for sending messages
+
+        Parameters:
+        -----------
+        handler : MessageHandler
+            A MessageHandler instance for the desired notification method.
+        """
+        self.message_handlers.append(handler)
+
+    def notify(self, message):
+        """Send a notification to all registered handlers by calling .send_message() """
+        for handler in self.message_handlers:
+            handler.send_message(message)
```

Comment:
**question**: What if you want to send a message to the console and log file by not send an email?

Reply:
hmm... I would assume you'd be relying on `logging` or `print()` statements to get info to the console. I add `ConsoleHandler` to the handlers in `__init__()` to ensure that any error messages picked up by `global_exception_handler` didn't get silently clobbered if no email is set up/isn't working. 

This isn't meant to be a wrapper for logging, but rather a point-in-time notification (after a certain chunk or all of a script is finished) that also ensures we get notified of any uncaught exceptions (except, of course, exceptions in the notification process itself). For example, auditor will still have all it's logging and -v `print()` statements, but at the end I'll build the summary message that normally gets printed to the cli and send that as a notification, along with the gzip'ed audit report logfile.

Is that what you were asking?

Reply:
Yes, I see where you are going now. I was thinking that it would replace the default logger. I'm good with this.